mixer/pkg/runtime2/config/ephemeral.go

```diff
@@ -0,0 +1,383 @@
+		log.Debugf("config.Snapshot contents:\n%s", s.String())
+	}
+	return s
+}
+
+func (e *Ephemeral) processAttributeManifests(counters Counters) map[string]*config.AttributeManifest_AttributeInfo {
+	if !e.attributesChanged && e.cachedAttributes != nil {
+		return e.cachedAttributes
+	}
+
+	attrs := make(map[string]*config.AttributeManifest_AttributeInfo)
+	for k, obj := range e.entries {
+		if k.Kind != AttributeManifestKind {
+			continue
+		}
+
+		log.Debug("Start processing attributes from changed manifest...")
+
+		cfg := obj.Spec
+		for an, at := range cfg.(*config.AttributeManifest).Attributes {
+			attrs[an] = at
+
+			if log.DebugEnabled() {
+				log.Debugf("Attribute '%s': '%s'.", an, at.ValueType.String())
+			}
+		}
+	}
+
+	// append all the well known attribute vocabulary from the templates.
+	//
+	// ATTRIBUTE_GENERATOR variety templates allows operators to write Attributes
+	// using the $out.<field Name> convention, where $out refers to the output object from the attribute generating adapter.
+	// The list of valid names for a given Template is available in the Template.Info.AttributeManifests object.
+	for _, info := range e.templates {
+		log.Debugf("Processing attributes from template: '%s'", info.Name)
+
+		for _, v := range info.AttributeManifests {
+			for an, at := range v.Attributes {
+				attrs[an] = at
+
+				if log.DebugEnabled() {
+					log.Debugf("Attribute '%s': '%s'", an, at.ValueType.String())
+				}
+			}
+		}
+	}
+
+	log.Debug("Completed processing attributes.")
+	counters.attributes.Add(float64(len(attrs)))
+
+	return attrs
+}
+
+func (e *Ephemeral) processHandlerConfigs(counters Counters) map[string]*Handler {
+	handlers := make(map[string]*Handler)
+
+	for key, resource := range e.entries {
+		var info *adapter.Info
+		var found bool
+		if info, found = e.adapters[key.Kind]; !found {
+			continue
+		}
+
+		adapterName := key.String()
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming handler config: name='%s'\n%s", adapterName, resource.Spec.String())
+		}
+
+		cfg := &Handler{
+			Name:    adapterName,
+			Adapter: info,
+			Params:  resource.Spec,
+		}
+
+		handlers[cfg.Name] = cfg
+	}
+
+	counters.handlerConfig.Add(float64(len(handlers)))
+	return handlers
+}
+
+func (e *Ephemeral) processInstanceConfigs(counters Counters) map[string]*Instance {
+	instances := make(map[string]*Instance)
+
+	for key, resource := range e.entries {
+		var info *template.Info
+		var found bool
+		if info, found = e.templates[key.Kind]; !found {
+			continue
+		}
+
+		instanceName := key.String()
+
+		if log.DebugEnabled() {
+			log.Debugf("Processing incoming instance config: name='%s'\n%s", instanceName, resource.Spec.String())
+		}
+
+		cfg := &Instance{
+			Name:     instanceName,
+			Template: info,
+			Params:   resource.Spec,
+		}
+
+		instances[cfg.Name] = cfg
+	}
+
+	counters.instanceConfig.Add(float64(len(instances)))
+	return instances
+}
+
+func (e *Ephemeral) processRuleConfigs(
```

Comment:
This is `controller.processRules()`, 
it would be easier if the purely structural changes are separated from performance changes.

Reply:
I've tried to ease the burden by sending two previous packs:
https://github.com/istio/istio/pull/2450
https://github.com/istio/istio/pull/2448
Perhaps starting with those would be easier.

Reply:
Ok, I will look at #2450